### PR TITLE
 semanal: Check that async for/with is inside an async function

### LINKED ISSUE #10594: async for/async with are allowed in non-async functions
**To Reproduce**

```
async def b():
 for i in range(5):
  yield i

def a():
 return [x async for x in b()]
```

**Expected Behavior**

Syntax error. When trying to run with python this code gives a syntax error.

**Actual Behavior**

Passed with no errors.

**Your Environment**

- Mypy version used: 0.812
- Python version used: 3.8

### PR DESCRIPTION
**Pull request by shoracek**
### Description

Added a check that async for/with is in an async function.
The code and error message are similar to those checking await.

Fixes #10594

## Test Plan

I used an expanded example from the issue as a unit test. The example was extended to also cover for loop, with expression and usage outside of a function.

Also fixed another test case using async with expression outside of an async function.

### CODE REVIEW

**Comment by jhance** [comment]
We should probably have an error code. Also, the format string for the message should be in the message registry.

**Comment by jhance** [comment]
Is there a reason the `ast` module itself doesn't raise for syntax?

**Comment by JelleZijlstra** [comment]
Some SyntaxErrors are raised after parsing in the Python compiler, e.g. invalid nonlocal declarations, some things with async. So mypy has to replicate such checks in semanal.

**Comment by shoracek** [comment]
I have kept the error code as `misc`, made the messages use terminology that the Python's syntax error uses, and added the message to the registry.

**Comment by jhance** [comment]
Yeah, syntax seems better if it already exists.

**Comment by shoracek** [comment]
Used `syntax` as the error code and also set it to be a blocking error, since Python will not run code with this error.

File: mypy/semanal.py [hunk 1]
```diff
@@ -4173,6 +4183,11 @@ def visit_dictionary_comprehension(self, expr: DictionaryComprehension) -> None:
         self.analyze_comp_for_2(expr)
 
     def visit_generator_expr(self, expr: GeneratorExpr) -> None:
+        if True in expr.is_async:
```
Comment (JelleZijlstra) [inline mypy/semanal.py:4186]: An async genexp is actually legal in a non-async function: `def f(): return (x async for x in y)`. Could you add a test for that?

**Comment by JelleZijlstra** [comment]
Not sure we should use a blocking error. There's a risk that this check has a bug or that a future Python version changes the rules, and in that case it would be useful for users to be able to skip this error. I don't think we should be using blocking errors for anything we can recover from.

**Comment by shoracek** [comment]
As suggested, I've switched it back to non-blocking error, removed the error for generator expression, and added the remaining expressions into the test.

**Comment by jhance** [comment]
It looks good. Can we change `True in ...` to `any(genr.is_async)` for all of the cases where we have something like that? Afterwards I think we can merge this.

**Comment by shoracek** [comment]
Done!

**Pull request merged**